### PR TITLE
Fix pruning for case with disconnected NNCFGraph

### PR DESCRIPTION
**Pull request by AlexanderDokuchaev**
If model used unregistered module (like DeformConv2d from mmcv.ops) NNCFgraph can be disconnected.
As result some node have no input or output edges. 

* Added calculation of output shape for node without output edge
* Added check for result of `get_previous_conv`
* identity_mask_propagation propagate `none` if no input mask

### CODE REVIEW

File: nncf/torch/pruning/filter_pruning/algo.py [hunk 1]
```diff
@@ -253,12 +256,36 @@ def _init_pruned_modules_params(self):
             self.pruning_quotas[cluster.id] = np.floor(self._modules_out_channels[cluster.elements[0].node_name] \
                                                        * self.pruning_quota)
 
+    def _calculate_output_shape(self, graph: NNCFGraph, node: NNCFNode) -> Tuple[int, ...]:
```
Comment (AlexKoff88) [inline nncf/torch/pruning/filter_pruning/algo.py:259]: Can we have a test for this case?

I think we will have more such cases in the future. Is it possible to introduce a logic that we check the connectivity of the NNCF graph and raise an exception (or maybe a warning) if there are problems?

File: nncf/torch/pruning/filter_pruning/algo.py [hunk 2]
```diff
@@ -253,12 +256,36 @@ def _init_pruned_modules_params(self):
...
+        """
+        Calculates output shape of convolution layer by input edge.
+        :param graph: the model graph
```
Comment (ljaljushkin) [inline nncf/torch/pruning/filter_pruning/algo.py:262]: to be aligned with CodeStyle guidelines.

```suggestion
        Calculates output shape of convolution layer by input edge.
        
        :param graph: the model graph
```

**Review by ljaljushkin — CHANGES_REQUESTED** [review]

File: nncf/common/pruning/mask_propagation.py [hunk 3]
```diff
@@ -77,6 +77,9 @@ def identity_mask_propagation(node: NNCFNode, graph: NNCFGraph):
     Propagates input mask through nx_node.
     """
     input_masks = get_input_masks(node, graph)
+    if not input_masks:
+        # In case for disconnected NNCFGraph
+        input_masks = [None]
```
[the remark this answers is not in the document]
Comment (AlexanderDokuchaev) [reply nncf/common/pruning/mask_propagation.py:82]: Before if  `input_masks == []` nncf crashed with assert, assert len(input_masks) == 1. 
This is not affect all supported models, only models with disconnected graph.

File: nncf/torch/pruning/filter_pruning/algo.py (hunk 1 above)
Comment (AlexanderDokuchaev) [reply nncf/torch/pruning/filter_pruning/algo.py:259]: Yes, will add tests.

I think add warning before use _calculate_output_shape, with message that node have output edge. ​
Exception was before on line `out_edge = graph.get_output_edges(node)[0]` with out of index.

> Is it possible to introduce a logic that we check the connectivity

It's possible, but we need discussion (what logic in nncf it can affect? what we can do with disconnected graph? list of test cases). I will create issue about it.

File: nncf/common/pruning/mask_propagation.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ljaljushkin) [reply nncf/common/pruning/mask_propagation.py:82]: How did you know? Did you check it manually? Don't think we should do it in that way further: it's time consuming and not reliable. 
IMHO, there should be automatic tests for `ModelPruner` and `MaskPropagationAlgorithm`, checking propagation of masks and covering corner cases like this.

File: nncf/torch/pruning/filter_pruning/algo.py (hunk 1 above)
Comment (AlexKoff88) [reply nncf/torch/pruning/filter_pruning/algo.py:259]: Besides creating an issue, can you also think about the proposal?

File: nncf/common/pruning/mask_propagation.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (AlexanderDokuchaev) [reply nncf/common/pruning/mask_propagation.py:82]: About tests, yes, i will add it.

> How did you know? Did you check it manually? Don't think we should do it in that way further: manual testing is time consuming and not reliable.

On current moment models with disconnected graph and empty input masks can not be pruned because nncf crash on `assert len(input_masks) == 1`.  So we have no supported pruned models that affect by this changes.

**Review by ljaljushkin — APPROVED** [review]

**Review by AlexKoff88 — APPROVED** [review]

**Comment by AlexKoff88** [comment]
@evgeniya-egupova, do you have any comments here?

File: nncf/torch/pruning/filter_pruning/algo.py [hunk 4]
```diff
@@ -253,12 +256,38 @@ def _init_pruned_modules_params(self):
...
+                out_edge = graph.get_output_edges(node)[0]
+                out_shape = out_edge.tensor_shape[2:]
+            else:
+                # For disconnected NNCFGraph when convolution layers have no output edge
+                out_shape = self._calculate_output_shape(graph, node)
+                nncf_logger.error("Node %s have no output edge in NNCFGraph", node.node_name)
+            self._modules_out_shapes[node.node_name] = out_shape
 
         for node in graph.get_nodes_by_types([v.op_func_name for v in NNCF_LINEAR_MODULES_DICT]):
             out_edge = graph.get_output_edges(node)[0]
```
Comment (evgeniya-egupova) [inline nncf/torch/pruning/filter_pruning/algo.py:293]: What if linear won't have output edge?

**Comment by evgeniya-egupova** [comment]
Also, can convolutions be missing an input edge? For example with unregistered activation module?

File: nncf/torch/pruning/filter_pruning/algo.py (hunk 4 above)
Comment (AlexanderDokuchaev) [reply nncf/torch/pruning/filter_pruning/algo.py:293]: Exception "out of index".
I'll add same logic for linear.

**Comment by AlexanderDokuchaev** [comment]
> Also, can convolutions be missing an input edge? For example with unregistered activation module?

In this case we can do nothing.

File: nncf/torch/pruning/filter_pruning/algo.py (hunk 4 above)
Comment (AlexanderDokuchaev) [reply nncf/torch/pruning/filter_pruning/algo.py]: done

**Comment by AlexanderDokuchaev** [comment]
In case of missing an input edge, we can not get information about W and H

**Comment by evgeniya-egupova** [comment]
> In case of missing an input edge, we can not get information about W and H

Actually, we can store input/output shapes as node attributes. This has been a topic of discussion before, and @vshampor insisted that we keep shapes in edges and find ways to fix the graph if problems occur. However, we now have at least two new cases where some nodes are missing input/output edges. I think we need to discuss this again.

**Comment by vshampor** [comment]
> > In case of missing an input edge, we can not get information about W and H
> 
> Actually, we can store input/output shapes as node attributes. This has been a topic of discussion before, and @vshampor insisted that we keep shapes in edges and find ways to fix the graph if problems occur. However, we now have at least two new cases where some nodes are missing input/output edges. I think we need to discuss this again.

All of this is fixed if we have a convention to add "constant", or "unknown" nodes into the NNCFGraph wherever we know input shape, but not the originating operation. This is in itself a workaround, however, and must be explicitly marked as such. This is better fixed by making sure all PT ops are registered and wrapped.

File: nncf/common/pruning/mask_propagation.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mkaglins) [reply nncf/common/pruning/mask_propagation.py:82]: Agreed with @ljaljushkin, all cases (include valid graphs) should be covered by tests.

**Review by mkaglins — APPROVED** [review]
